src/communication/securechannel.cpp

```diff
@@ -0,0 +1,301 @@
+
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <memory>
+#include <sstream>
+
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/ssl.h>
+#include <openssl/trace.h>
+
+#include <utils/cryptohelper.hpp>
+#include <utils/pkcs11helper.hpp>
+
+#include "logger/logmodule.hpp"
+#include "securechannel.hpp"
+
+namespace aos::mp::communication {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+SecureChannel::SecureChannel(const aos::mp::config::Config& cfg, CommChannelItf& channel,
+    aos::mp::iamclient::CertProviderItf& certProvider, aos::cryptoutils::CertLoaderItf& certLoader,
+    aos::crypto::x509::ProviderItf& cryptoProvider, int port, const std::string& certStorage)
+    : mChannel(&channel)
+    , mCertProvider(&certProvider)
+    , mCertLoader(&certLoader)
+    , mCryptoProvider(&cryptoProvider)
+    , mCfg(&cfg)
+    , mPort(port)
+    , mCertStorage(certStorage)
+{
+    LOG_DBG() << "Create secure channel: port=" << mPort;
+
+    InitOpenssl();
+
+    const SSL_METHOD* method = TLS_server_method();
+    mCtx                     = CreateSSLContext(method);
+
+    ENGINE* eng = ENGINE_by_id("pkcs11");
+    if (!eng) {
+        throw std::runtime_error("Failed to load PKCS11 engine");
```

Comment:
Looks like SSL context will not be  released in case of exceptions.
As per current implementation, application will not start, but should we improve resource handling for this scenario?

Reply:
Done